[compiler-rt] Realtime Sanitizer: Introduce Realtime Sanitizer (RTSan) backend

cjappl opened the pull request:
Introducing the main runtime of realtime sanitizer. For more information, please see the [discourse thread](https://discourse.llvm.org/t/rfc-nolock-and-noalloc-attributes/76837). We have also put together a [reviewer support document](https://github.com/realtime-sanitizer/radsan/blob/doc/review-support/doc/review.md) to show what our intention is.

This review introduces the sanitizer backend. This includes:
* CMake build files (largely adapted from asan).
* Main RTSan architecture (the external API, thread local context, stack).
* Interceptors.
* Many unit tests.

Please see the [reviewer support document](https://github.com/realtime-sanitizer/radsan/blob/doc/review-support/doc/review.md) for what our next steps are. We are moving in lockstep with this PR #84983 for the codegen coming up next.

Note to reviewers: If you see support documentation mention "RADSan", this was the "old acronym" for the realtime sanitizer, they refer to the same thing. If you see it let us know and we can correct it (especially in the llvm codebase)

cjappl committed:
First try

cjappl committed:
Fixing some clang tidy foo

cjappl committed:
One round of renaming

cjappl committed:
Another round of style cleanups

cjappl committed:
more auto removal, more case fixing

cjappl committed:
Missing end of namespace comment

cjappl committed:
Fix one naming issue

cjappl committed:
No more unistd

cjappl committed:
Radsan context moved to impl

cjappl committed:
Promise lit tests later

cjappl committed:
Move radsan_test and add helper comment to top of file

cjappl committed:
Only link radsan statically

cjappl committed:
Fix comments in radsan.h

cjappl committed:
Codeowners

davidtrevelyan committed:
Replace inlined exit with selected action stub

cjappl committed:
Fix python formatting errors

cjappl committed:
Fix cpp style issues

cjappl commented on clang/lib/Driver/ToolChains/CommonArgs.cpp:
```diff
@@ -1382,6 +1382,10 @@ collectSanitizerRuntimes(const ToolChain &TC, const ArgList &Args,
 
+  if (!SanArgs.needsSharedRt() && SanArgs.needsRadsanRt()) {
+    StaticRuntimes.push_back("radsan");
+  }
```
One question we have here: What is the SharedRuntimes vs StaticRuntimes? Why would we need to support both? This is one change we "eyeballed" based on the other sanitizers.

MaskRay replied:
I have some notes at https://maskray.me/blog/2023-01-08-all-about-sanitizer-interceptors#static-runtime

Nit: we omit braces in this single-line statement case https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

cjappl commented on compiler-rt/cmake/Modules/AllSupportedArchDefs.cmake:
```diff
@@ -32,6 +32,9 @@ set(ALL_ASAN_SUPPORTED_ARCH ${X86} ${X86_64} ${ARM32} ${ARM64} ${RISCV64}
 set(ALL_DFSAN_SUPPORTED_ARCH ${X86_64} ${MIPS64} ${ARM64} ${LOONGARCH64})
+set(ALL_RADSAN_SUPPORTED_ARCH ${X86} ${X86_64} ${ARM32} ${ARM64} ${RISCV64}
+    ${MIPS32} ${MIPS64} ${PPC64} ${S390X} ${SPARC} ${SPARCV9} ${HEXAGON}
+    ${LOONGARCH64})
```
This was copied from ASAN, is this acceptable? We don't have all architectures to test, although we think that our interceptors are simple enough to work on any machine (no platform specific assembly in our code, for example)

cjappl commented:
Pinging possibly interested parties for review. 

@yln @vitalybuka @Sirraide @AaronBallman @dougsonos @davidtrevelyan 

Please feel free to add more, we don't know who all may be interested. Thanks in advance

MaskRay commented on compiler-rt/lib/radsan/radsan.h:
```diff
@@ -0,0 +1,50 @@
+  violations are detected. Calls to this method are injected at the code
+  generation stage when RADSan is enabled.
+*/
+SANITIZER_INTERFACE_ATTRIBUTE void radsan_realtime_enter();
```
Use `__*` names to avoid conflicts with user code. Use `//`.

If there are user APIs, define them at `include/sanitizer/*_interface.h`

cjappl committed:
PR: Delete python header comment

cjappl commented on compiler-rt/test/radsan/lit.cfg.py:
```diff
@@ -0,0 +1,71 @@
+config.name = "RADSAN" + config.name_suffix
+
+# Setup source root.
+config.test_source_root = os.path.dirname(__file__)
```
Yes, the lit tests may be placed directly in `compiler-rt/test/radsan/` and they run as intended. Just checked on our branch that has them implemented.

cjappl committed:
PR: Change main namespace name to __radsan

cjappl committed:
PR: Fix stack trace namespace

cjappl committed:
PR: Fix incorrect C++ style comments

cjappl committed:
PR: Prefix external API with dunderscores

cjappl commented on compiler-rt/lib/radsan/radsan_context.cpp:
```diff
@@ -0,0 +1,106 @@
+
+namespace detail {
+
+static pthread_key_t Key;
```
Great thanks!! Before I go through and edit these, `snake_case` is preferred for variables in all contexts?

Really appreciate the review. Pushing up most of your requested changes now.

MaskRay replied:
Yes, `variable_name` and `FunctionName`.. Sorry if this change may take a while.
asan/hwasan/lsan are good examples to follow.

cjappl replied:
No apology necessary! Happy to do this right the first time.

Does this also apply to lambdas?

```
  auto Func = [&vec]() { vec.push_back(0.4f); };
```

Should this become:
```
  auto func = [&vec]() { vec.push_back(0.4f); };
```

Or remain "function case"?

cjappl committed:
PR: Fix variable naming cases for variables, leave lambdas for now

cjappl committed:
PR: Fix style nit

cjappl committed:
PR: Fix small style fallout

MaskRay commented:
https://discourse.llvm.org/t/rfc-nolock-and-noalloc-attributes/76837/75 contains a discussion about the subproject name.
Do you have opinions? @llvm/pr-subscribers-sanitizer @compnerd @petrhosek  @isanbard

davidtrevelyan committed:
Update comment and TU-local function style

cjappl committed:
PR: Remove too early inclusion of sanitizer common lit tests

cjappl commented:
Weekly ping of reviewers

@yln @vitalybuka @Sirraide @AaronBallman @zygoloid @compnerd @petrhosek @isanbard

Looking for weighing in on any of the code as it stands now, or the naming question posed by MaskRay [here](https://discourse.llvm.org/t/rfc-nolock-and-noalloc-attributes/76837/75), so far there seems to be support for rtsan over radsan :)

cjappl commented:
Ping

@yln @vitalybuka @Sirraide @AaronBallman @zygoloid @compnerd @petrhosek @isanbard @MaskRay

dougsonos commented:
> Looking for weighing in on any of the code as it stands now, or the naming question posed by MaskRay [here](https://discourse.llvm.org/t/rfc-nolock-and-noalloc-attributes/76837/75), so far there seems to be support for rtsan over radsan :)

+1 for rtsan

cjappl committed:
Test failure: fix initialization of pthread_cond_t

cjappl committed:
Test failure: Try again with initialization of pthread_cond_t

cjappl committed:
Fix code formatting on unit test

cjappl committed:
Ensure all pthread_cond_t are initialized in unit tests

cjappl committed:
Fix issue where CREAT flag was not properly passing mode (#23)

* Pass on CREAT mode propertly in open interceptors

* Add unit tests

* Remove two unneccessary bits

* Fix fcntl

cjappl committed:
Fix issue where fcntl was not using last arg correctly (#24)

* Fix issue where fcntl was not using last arg correctly

* Comment update

cjappl committed:
Introduce test fixture for temporary files (#25)

* Introduce test fixture for temporary files

* Style fixup for method name

davidtrevelyan committed:
Clean out temp files in test fixture SetUp too

davidtrevelyan committed:
Rename RADSan to RTSan

vitalybuka commented on clang/lib/Driver/ToolChains/Darwin.cpp:
```diff
@@ -1509,6 +1511,11 @@ void DarwinClang::AddLinkRuntimeLibArgs(const ArgList &Args,
         AddLinkSanitizerLibArgs(Args, CmdArgs, "asan");
       }
```
Driver needs tests

cjappl replied:
Thanks for catching this! Added a few basic ones that test that the flag is supported, and that some known sanitizer incompatibilities are blocked

cjappl committed:
First working rtsan driver tests

cjappl committed:
Add linux working test

cjappl committed:
Another test for openBSD

cjappl committed:
Add feature for realtime sanitizer, add incompatible sanitizers

cjappl committed:
More incompatabilities added

cjappl committed:
rtsan tests to fsanitize

cjappl committed:
Deleted rtsan.c

cjappl committed:
Code formatting

cjappl commented:
Bumped the branch to resolve the conflict in config-ix.cmake

cjappl committed:
Fix python styling

cjappl commented:
Pinging reviewers @vitalybuka @MaskRay @zygoloid 

All the comments on this PR have been addressed, looking for more feedback, or approval/merge if we are getting close! Thanks in advance :)

MaskRay commented:
> This is PR 1/?? (3? 4?) introducing the main guts of the realtime sanitizer. For more information, please see the [discourse thread](https://discourse.llvm.org/t/rfc-nolock-and-noalloc-attributes/76837). We have also put together a [reviewer support document](https://github.com/realtime-sanitizer/radsan/blob/doc/review-support/doc/review.md?rgh-link-date=2024-05-16T21%3A34%3A46Z) to show what our intention is.

clangDriver changes are usually the last. The expectation is that if `-fsanitize=realtime` does not return an error, there should be some basic functionality.

* clangCodeGen , testing with `%clang_cc1`
* clangDriver, testing with `%clang`

We typically split clangDriver and compiler-rt changes. Runtime compiler-rt tests often expose minor issues that only fail on certain platforms. Keeping it separate prevents us from potentially reverting the compiler-rt code, leading to unnecessary churn.
The recent numerical stability sanitizer did not do a good job for this aspect. I apologize for that as a reviewer.)

cjappl committed:
PR: Get rid of clang components only leave compiler-rt

cjappl commented:
> clangDriver changes are usually the last. The expectation is that if -fsanitize=realtime does not return an error, there should be some basic functionality.

Thanks for the feedback @MaskRay . I removed all of the clang components of this review, leaving only the compiler-rt components. We will re-introduce the clang changes in a future review.

In the meantime, I commented out compilation of our unit tests that required the -fsanitize flag, let me know if you'd prefer to see them removed until later. These unit tests are paired with the code introduced here, so I think at least by subject it makes sense to introduce them here, even if they are briefly not being compiled.

MaskRay commented on compiler-rt/lib/rtsan/tests/rtsan_test_functional.cpp:
```diff
@@ -0,0 +1,207 @@
+using namespace std::chrono_literals;
+
+TEST(TestRtsan, VectorPushBackAllocationDiesWhenRealtime) {
+  std::vector<float> vec{};
```
omit `{};`

MaskRay commented on compiler-rt/test/rtsan/lit.cfg.py:
```diff
@@ -0,0 +1,67 @@
+# other platforms.
+if config.libcxx_used == "1" and config.host_os == "Linux":
+    clang_rtsan_cflags = clang_rtsan_cflags + (
+        ["-L%s -lc++ -Wl,-rpath=%s" % (config.llvm_shlib_dir, config.llvm_shlib_dir)]
```
compiler-rt/test/xray/lit.cfg.py uses this construct but other sanitizers don't. Is it needed?

cjappl replied:
Removed!

MaskRay commented on compiler-rt/test/rtsan/lit.cfg.py:
```diff
@@ -0,0 +1,67 @@
+        "%rtsanlib",
+        (
+            "-lm -lpthread %s -lrt -L%s "
+            "-Wl,-whole-archive -lclang_rt.rtsan%s -Wl,-no-whole-archive"
```
clangdriver shall pass these arguments to the linker. lit.cfg.py should not need to do it.

MaskRay commented on compiler-rt/lib/rtsan/rtsan_context.cpp:
```diff
@@ -0,0 +1,97 @@
+static pthread_key_t context_key;
+static pthread_once_t key_once = PTHREAD_ONCE_INIT;
+
+static void internalFree(void *ptr) { __sanitizer::InternalFree(ptr); }
```
`InternalFree`. If there is a name conflict, perhaps pick a different name.

MaskRay commented on compiler-rt/lib/rtsan/rtsan_context.cpp:
```diff
@@ -0,0 +1,97 @@
+*/
+static void InvokeViolationDetectedAction() { exit(EXIT_FAILURE); }
+
+namespace __rtsan {
```
https://llvm.org/docs/CodingStandards.html#use-namespace-qualifiers-to-implement-previously-declared-functions

If `using namespace __rtsan`, `.cpp` files should not need `namespace __rtsan`.

MaskRay commented on compiler-rt/lib/rtsan/rtsan_stack.cpp:
```diff
@@ -0,0 +1,50 @@
+  OverrideCommonFlags(cf);
+}
+
+using namespace __rtsan;
```
should be moved up, beside `using namespace __sanitizer;`

MaskRay commented on compiler-rt/lib/rtsan/rtsan_interceptors.cpp:
```diff
@@ -0,0 +1,412 @@
+#include "interception/interception.h"
+#include "rtsan/rtsan_context.h"
+
+#if !SANITIZER_LINUX && !SANITIZER_APPLE
```
If other OSes are not supported, the NetBSD/FreeBSD/etc references in `lit.cfg.py` should be removed.

MaskRay reviewed (commented):
Sorry for the delay. I have a really huge code review backlog...

MaskRay commented:
> In the meantime, I commented out compilation of our unit tests that required the -fsanitize flag, let me know if you'd prefer to see them removed until later. These unit tests are paired with the code introduced here, so I think at least by subject it makes sense to introduce them here, even if they are briefly not being compiled.

Yeah, disabling the testsuite should be safe. The tests can be enabled after clangCodeGen and clangDriver patches landed.

cjappl committed:
PR: MaskRay comments

- remove empty initializers
- better namespace compliance with LLVM standards
- Fix lit.cfg.py to remove crufty elements not needed
- Change name of InternalFree, add explainer comment

cjappl commented:
Pinging reviewers @vitalybuka @zygoloid

State of the review: 1 approval from MaskRay, all open comments have been addressed. Looking for more feedback, or more approval or a merge if this is looking good.

Thanks much! :)

vitalybuka commented on compiler-rt/CODE_OWNERS.TXT:
```diff
@@ -67,3 +67,7 @@ D: ThreadSanitizer
 N: Bill Wendling
 E: isanbard@gmail.com
 D: Profile runtime library
+
```
BTW @alexander-shaposhnikov NSAN is missing from here.

vitalybuka commented on compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp:
```diff
@@ -0,0 +1,516 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
```
In my experience LIT tests in are easier to maintain in LLVM environment

`TEST(TestRtsan,` are internals test, so probably fine to keep as GTEST

However `TEST(TestRtsanInterceptors` are likely better as regular LIT tests

vitalybuka replied:
Actually even TestRtsan mostly can be done as LIT

BTW. Why I don't like GTEST in LLVM?

LIT is easier to reproduce and experiment and debug than with a particular test case of.

Also I don't trust incremental builds with GTEST, some times I am not sure that my changes are used in tests.

A couple of time CMakes bugs caused compiler-rt GTEST not executed on bots and it was unnoticed for a long time. LIT is majority, so unlikely it could happen with them.

cjappl replied:
Seems reasonable to convert these to lit! I propose leaving this for this commit, and we convert them in the near future. Specifically this next commit will introduce the -fsanitize=realtime flag, and come with a bunch of the lit tests that we have written, so it will be easier to convert them once that infrastructure has landed.

One clarifying question, for when that work is done:

Should we have a lit test for every intereceptor? The reason we did it in gtest unit tests is that adding a new interceptor's test is a few lines, as opposed to a  brand new file with the lit boilerplate. This makes it easy to make sure that every interceptor does what we think it does. We worried about the "cost" of the new file for every interceptor.  In our demo branch, as we have it laid out, the lit tests test on the full end-to-end system level, while these tests actually hit every interceptor.

vitalybuka replied:
> Should we have a lit test for every intereceptor? 

I don't have strong preference. For other sanitizer default interceptor behavior is to crash, so we usually create a test per interceptor.

> The reason we did it in gtest unit tests is that adding a new interceptor's test is a few lines, as opposed to a brand new file with the lit boilerplate. This makes it easy to make sure that every interceptor does what we think it does. We worried about the "cost" of the new file for every interceptor. In our demo branch, as we have it laid out, the lit tests test on the full end-to-end system level, while these tests actually hit every interceptor.

Yes, boilerplate of LIT is not nice. So there is a trade off. I guess you can wait and see.

vitalybuka reviewed (approved):
LGTM. but I suggest to rework test into LIT, before or either commit.

cjappl commented:
Also, just as a heads up, neither of the authors have commit access as we are new to LLVM. If this is OK to go in we will need help with that step.

vitalybuka commented:
I can reproduce and will debug in an hour.

cjappl commented:
> I can reproduce and will debug in an hour.

Thanks for the help @vitalybuka , I am looking into this as well, but very grateful for more experienced eyes on the matter too. I'll keep you posted here if I find anything

cjappl commented:
There are 3 errors I see:

Mostly it's this `llvm_gtest` issue:
```
ninja: error: 'compiler-rt/lib/rtsan/tests/llvm_gtest', needed by 'compiler-rt/lib/rtsan/tests/RtsanNoInstTestObjects.rtsan_preinit.cpp.x86_64.o', missing and no known rule to make it
```

On fuchsia it is a parameter type issue:
```
/var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/compiler-rt/lib/rtsan/rtsan_interceptors.cpp:55:22: error: use of undeclared identifier '__unsanitized_open'; did you mean '__unsanitized_memset'?
   55 |   const int result = REAL(open)(path, oflag, mode);
      |                      ^~~~~~~~~~
      |                      __unsanitized_memset
/var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/compiler-rt/lib/rtsan/../interception/interception.h:241:18: note: expanded from macro 'REAL'
  241 | # define REAL(x) __unsanitized_##x
      |                  ^
<scratch space>:138:1: note: expanded from here
  138 | __unsanitized_open
      | ^
/usr/local/fuchsia/sdk/arch/x64/sysroot/include/zircon/sanitizer.h:34:18: note: '__unsanitized_memset' declared here
   34 | __typeof(memset) __unsanitized_memset;
      |                  ^
/var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/compiler-rt/lib/rtsan/rtsan_interceptors.cpp:55:33: error: cannot initialize a parameter of type 'void *' with an lvalue of type 'const char *'
   55 |   const int result = REAL(open)(path, oflag, mode);
      |                                 ^~~~
/usr/local/fuchsia/sdk/arch/x64/sysroot/include/zircon/sanitizer.h:34:18: note: passing argument to parameter here
   34 | __typeof(memset) __unsanitized_memset;
      |                  ^
/var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/compiler-rt/lib/rtsan/rtsan_interceptors.cpp:69:22: error: use of undeclared identifier '__unsanitized_openat'
   69 |   const int result = REAL(openat)(fd, path, oflag, mode);
      |                      ^
/var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/compiler-rt/lib/rtsan/../interception/interception.h:241:18: note: expanded from macro 'REAL'
  241 | # define REAL(x) __unsanitized_##x
      |                  ^
<scratch space>:141:1: note: expanded from here
  141 | __unsanitized_openat
      | ^
/var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/compiler-rt/lib/rtsan/rtsan_interceptors.cpp:77:22: error: use of undeclared identifier '__unsanitized_creat'
   77 |   const int result = REAL(creat)(path, mode);
      |                      ^
/var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/compiler-rt/lib/rtsan/../interception/interception.h:241:18: note: expanded from macro 'REAL'
  241 | # define REAL(x) __unsanitized_##x
      |                  ^
<scratch space>:144:1: note: expanded from here
  144 | __unsanitized_creat
      | ^
Step 6 (build) failure: build (failure)
...
```

On windows (which I think we should just disable support for, for now:

```
C:\b\slave\sanitizer-windows\llvm-project\compiler-rt\lib\rtsan\rtsan_context.cpp(19): fatal error C1083: Cannot open include file: 'pthread.h': No such file or directory
```


Working to see if I can fix the llvm_gtest issue right now. I would love advice on Fuchsia and disabling rtsan for windows.

cjappl commented:
This patch I have a strong hunch fixes issue 1 in my comment above. This patch compiles and runs on my linux container and my mac, proving it doesn't affect compilation on those two systems.

```
diff --git a/compiler-rt/lib/rtsan/tests/CMakeLists.txt b/compiler-rt/lib/rtsan/tests/CMakeLists.txt
index d96e538b255f..e539ebc387e0 100644
--- a/compiler-rt/lib/rtsan/tests/CMakeLists.txt
+++ b/compiler-rt/lib/rtsan/tests/CMakeLists.txt
@@ -60,7 +60,7 @@ foreach(arch ${RTSAN_TEST_ARCH})
   #  RtsanUnitTests "Rtsan-${arch}-Test" ${arch}
   #  COMPILE_DEPS ${RTSAN_UNITTEST_HEADERS}
   #  SOURCES ${RTSAN_INST_TEST_SOURCES} ${COMPILER_RT_GOOGLETEST_SOURCES}
-  #  DEPS llvm_gtest rtsan
+  #  DEPS rtsan
   #  CFLAGS ${RTSAN_UNITTEST_CFLAGS} -fsanitize=realtime
   #  LINK_FLAGS ${RTSAN_UNITTEST_LINK_FLAGS} -fsanitize=realtime)

@@ -94,7 +94,6 @@ foreach(arch ${RTSAN_TEST_ARCH})
     COMPILE_DEPS ${RTSAN_UNITTEST_HEADERS}
     SOURCES ${RTSAN_NOINST_TEST_SOURCES}
             ${COMPILER_RT_GOOGLETEST_SOURCES}
-    DEPS llvm_gtest
     CFLAGS ${RTSAN_UNITTEST_CFLAGS}
     LINK_FLAGS ${RTSAN_UNITTEST_LINK_FLAGS}
     RUNTIME ${RTSAN_TEST_RUNTIME})
```

This also better matches the other sanitizers, llvm_gtest is a dependency in llvm, anything in compiler-rt using gtest just uses the sources directly as seen in:

```
SOURCES ${RTSAN_NOINST_TEST_SOURCES}
             ${COMPILER_RT_GOOGLETEST_SOURCES}
```

glandium commented:
This also breaks building the compiler-rt for older android versions, because pthread_spinlock_t is only defined when __ANDROID_API__ >= 24.

cjappl commented:
> This also breaks building the compiler-rt for older android versions, because pthread_spinlock_t is only defined when `__ANDROID_API__` >= 24.

Thanks for reporting this @glandium ! I am in the process of disabling this for android as we speak. Let me know if you see anything else fishy
